### ISSUE TITLE
Document the design of All

### ISSUE THREAD

**Issue by treeowl** [well-typed/generics-sop#137]
It's not at all obvious to the casual reader why `All` is defined as it is, and in particular what `SListI` is really about. This should be explained.

**Comment by kosmikus** [comment]
You're absolutely right, this should be documented. But even I have to recover that information, so it's not trivial. I'll try to do this.

**Comment by kosmikus** [comment]
(Also, the design has changed a number of times, and while `SListI` is now just `All Top`, it used to be separate. So in a way, the justification for even having `SListI` separately is now relatively weak.)
